lib/model/indexsorter.go

```diff
@@ -0,0 +1,224 @@
+}
+
+func (s *inMemoryIndexSorter) Close() {
+	// Nothing to clean up here.
+}
+
+// A leveldbIndexSorter inserts items keyed on LocalVersion, then iterates
+// over that when asked for batches.
+type leveldbIndexSorter struct {
+	db      *leveldb.DB
+	dbPath  string
+	prevKey int64
+}
+
+func newLeveldbIndexSorter() (*leveldbIndexSorter, error) {
+	path, err := ioutil.TempDir("", "indexCache")
+	if err != nil {
+		return nil, err
+	}
+	db, err := leveldb.OpenFile(path, &opt.Options{
+		// We're only going to read it once, so we don't need caching
+		DisableBlockCache: true,
+		// It should certainly not already exist
+		ErrorIfExist: true,
+		// We don't need sync writes on it, the data is worthless if we crash
+		NoSync: true,
+		// And we don't want to spend too many open files on it
+		OpenFilesCacheCapacity: 16,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	l.Debugln("leveldbIndexSorter at", path)
+	return &leveldbIndexSorter{
+		db:     db,
+		dbPath: path,
+	}, nil
+}
+
+func (s *leveldbIndexSorter) Enqueue(items ...protocol.FileInfo) {
+	// Marshal each item, keyed by it's LocalVersion.
+	for _, f := range items {
+		var key [8]byte
+		binary.BigEndian.PutUint64(key[:], uint64(f.LocalVersion))
+		if err := s.db.Put(key[:], f.MustMarshalXDR(), nil); err != nil {
+			panic(err)
+		}
+	}
+}
+
+func (s *leveldbIndexSorter) Batch() []protocol.FileInfo {
+	iter := s.db.NewIterator(nil, nil)
+
+	// We start iterating at the previous key plus one, or we'd get the last
+	// item once more. (They key not existing isn't an issue, the iterator
+	// will point at the next higher available one.)
+
+	var key [8]byte
+	binary.BigEndian.PutUint64(key[:], uint64(s.prevKey+1))
+
+	// Create a batch fulfilling the usual criteria.
+
+	var batch []protocol.FileInfo
+	batchSize := 0
+	for ok := iter.Seek(key[:]); ok && batchSize < batchTargetSize && len(batch) < batchMaxFiles; ok = iter.Next() {
+		var f protocol.FileInfo
+		if err := f.UnmarshalXDR(iter.Value()); err != nil {
+			panic(err)
+		}
+		batch = append(batch, f)
+		batchSize += sizeBytes(f)
+		s.prevKey = f.LocalVersion
+	}
+
+	return batch
+}
+
+func (s *leveldbIndexSorter) Close() {
+	s.db.Close()
+	os.RemoveAll(s.dbPath)
+}
+
+// --- Utility functions
+
+type sortByLocalVersion []protocol.FileInfo
+
+func (s sortByLocalVersion) Len() int {
+	return len(s)
+}
+func (s sortByLocalVersion) Swap(a, b int) {
+	s[a], s[b] = s[b], s[a]
+}
+func (s sortByLocalVersion) Less(a, b int) bool {
+	return s[a].LocalVersion < s[b].LocalVersion
+}
+
+// The approximate site of a FileInfo in memory.
+func sizeBytes(f protocol.FileInfo) int {
+	/*
+		type FileInfo struct {
+			Name         string // max:8192
+			Flags        uint32
+			Modified     int64
+			Version      Vector
+			LocalVersion int64
+			CachedSize   int64       // noencode (cache only)
+			Blocks       []BlockInfo // max:1000000
+		}
+	*/
+	return 8 + len(f.Name) + 4 + 8 + 8 + 16*len(f.Version) + 8 + 8 + 8 + 8 + 40*len(f.Blocks)
```

Comment:
I suggest you use unsafe.SizeOf() for the struct, and check that it matches the expected value, and panic if it doesn't, then add name, blocks and versions separately.